Quorum/apis/block_explorers/source_code.py

```diff
@@ -1,8 +1,19 @@
-from solidity_parser import parser
+import solcx
 from dataclasses import dataclass
+from pathlib import Path
+from enum import StrEnum
 
 import Quorum.utils.pretty_printer as pp
 
+# Install the latest version of Solidity compiler.
+solc_version = solcx.get_compilable_solc_versions()[0]
+solcx.install_solc(solc_version)
+solcx.set_solc_version(solc_version)
```

Comment:
This should be added to the README as well.
Also thinking if we should make it mandatory or optional+raising an error

Reply:
Whats the point to make it optional? the tool will crash.
What exactly you want in the readme? a line that says we use solc compiler to parse the contract and create AST?

Reply:
First, if we're installing solc compiler on user machine, he should know about it.
Second, not sure how solcx working, but if the user already have some solc installed on his machine, we do not want to change that...
And last, if the contract proposal is not compatible with the latest solc version, this might fail.

Reply:
Not sure if its possible to proposal not be compatible with the latest solc, the solcx installation is separated from the solc user install on his machine.

Reply:
Yes, Solidity contracts can indeed become incompatible with the latest Solidity compiler (`solc`). This can happen for several reasons:

1. **Breaking Changes in Solidity Versions**: Solidity often introduces new syntax, removes deprecated features, or changes existing functionalities. If a contract was written in an older version of Solidity, newer versions may introduce changes that render the contract incompatible. The Solidity compiler has a well-documented [Changelog](https://github.com/ethereum/solidity/blob/develop/Changelog.md) which lists these breaking changes by version.

2. **Version-Specific Features**: Some features, functions, or optimizations are specific to particular Solidity versions. If a contract uses features that were later modified or removed, it may need to be updated to compile under a newer `solc`.

3. **Pragma Directive Conflicts**: Contracts specify their compatible compiler version(s) using the `pragma solidity` directive (e.g., `pragma solidity ^0.8.0;`). If the specified range does not include the latest compiler, `solc` will refuse to compile it unless the `pragma` is updated or overridden.

4. **Library and Dependency Updates**: Some contracts depend on external libraries or modules that might also have version dependencies. If those libraries require specific Solidity versions, updating `solc` might introduce incompatibilities.

To ensure compatibility, it’s best to lock the contract to a specific Solidity version if it’s been thoroughly tested, or regularly update the contract to match the latest `solc` version if you want to take advantage of new features and improvements.

Reply:
I might be wrong here, but afaik latest solc isnt the best solution always, please verify this with @MichaelMorami

Reply:
Nice GPT answear, i know in general they can be but not in Proposals. as Proposals are **NEW** contracts.

Reply:
If we see complains about it we can always install the specific version mentioned in the source code.
But it's a super overkill solution which I am 99% sure is not necessary or relevant.

Reply:
And again i am stop on the parsing step, not on the compilation.

Reply:
I do agree with you in general, just trying to avoid future issues, and overkilling this since I lack knowledge but know for sure that a contract can be fixed on a constant version, even if it is a new contract.

Reply:
Your call i told my opinion about it, just warns you it will make the toll slower and blasts out the user computer with many solc versions

Reply:
For non visible gain.

Reply:
The solc used to compile the proposals doesn't have to be the latest available solc.
The contract pragma also doesn't have to specify a concrete version (^, >,<=, etc.).

Here's an example:
The latest proposal to date, No. 196, is compiled using solc 8.20, while the latest version out there is 8.28 if I'm not mistaken.
https://etherscan.io/address/0x18F74428230cB84c6FFF694614a453b9cE5aeb20
You can see in the bottom left corner of the screen shot the compiler version used. This is, AFAIK, necessary for the explorer in order to verify the contract based on the bytecode. This means that the compiler version it display is correct with high confidence. If it wasn't the code wouldn't have been verified
<img width="1708" alt="Screenshot 2024-11-04 at 11 32 18" src="https://github.com/user-attachments/assets/d9c35681-a192-48d7-8c6d-896517a6e02e">

Reply:
It's not that interesting as I use the solc to only parse the source code. so unless there is a syntax breakable change between solc which I think never occurs. will not encounter any issues with it.

Reply:
Until now we parsed it using Python solidity parser which has not been maintained for years. and we only encountered one contract that we couldn't parse with it. I am pretty sure of the robustness of the current solution.

Reply:
IMO, we can keep that solution for now, and if we encounter such an issue again we can fix it in the future by installing the concrete solc per contract solution.